### PR TITLE
Undo smooth

### PR DESCRIPTION
**Pull request by TurboGit**

### CODE REVIEW

**Comment by TurboGit** [comment]
This is one solution for issue described in #1418

**Comment by TurboGit** [comment]
This is a proposal and to initiate discussion about what I want to fix. Maybe there is better solution.

**Comment by houz** [comment]
I see some problems with the use of timestamps as done here – unless I am misunderstanding the code, I only looked over it once.
You are discarding undo entries if they have the same tag as the last one in the list. What if the undo step was important? Think about adding undo for image rating. The user rates an image A to 2 stars, then within the same time slot (thus with the same tag) he rates another image B as 3 stars. The undo list would discard the latter. So when pressing ctrl-z only image A would be reset. Of course the rating-undo could always use tag 0, but still, it feels like a race condition.
Instead of these tags I would suggest something like this: Add a timestamp field to `dt_undo_item_t` so that every entry has that. Set it to `dt_get_wtime()` when recording the undo step. Record every step. Then when undoing something take the latest undo item, remember its timestamp and type and undo everything from the top of the list until you reach something with a different type or that is more than SOME_TIME_PERIOD older than the first entry. That way you can guarantee that everything gets reverted, you no longer need to worry about creating meaningful tags in the users of the undo api.

At least that's my suggestion after looking at the PR. 😄

**Comment by TurboGit** [comment]
Your analysis is not quite correct as the recording is not done only on the timestamp but also on some id (e.g. mask id). But this imply a distributed complexity as it imposes some coding from all user of the undo feature.

Now, I like your proposal better as it avoid the distributed complexity as all the computation of what should be undone will be in the undo module itself. Lot better I would say... I'll take this route and propose a patch ASAP.

Thanks for the review houz.

**Comment by TurboGit** [comment]
Your analysis is not quite correct as the recording is not done only on the timestamp but also on some id (e.g. mask id). But this imply a distributed complexity as it imposes some coding from all user of the undo feature.

Now, I like your proposal better as it avoid the distributed complexity as all the computation of what should be undone will be in the undo module itself. Lot better I would say... I'll take this route and propose a patch ASAP.

Thanks for the review houz.

**Comment by TurboGit** [comment]
Likewise, if there is no more objection I'll merge at the end of the week.

File: src/views/undo.c [hunk 1]
```diff
@@ -19,12 +19,23 @@
 #include "views/undo.h"
 #include <glib.h>    // for GList, gpointer, g_list_first, g_list_prepend
 #include <stdlib.h>  // for NULL, malloc, free
+#include <sys/time.h>
+
+const unsigned long long MAX_TIME_PERIOD = 500; // 500ms
+
+static unsigned long long _get_timestamp(void)
```
Comment (houz) [inline src/views/undo.c]: Why are you not using `dt_get_wtime()`?

Comment (TurboGit) [inline src/views/undo.c]: Hum... embarrassing :) I did not notice this routine which seems to be what I need. I'll fix.

**Pull request merged**